cli/firmware/getversion.go

```diff
@@ -0,0 +1,136 @@
+func NewGetVersionCommand() *cobra.Command {
+
+	command := &cobra.Command{
+		Use:   "get-version",
+		Short: "Gets the version of the firmware the board is using.",
+		Long:  "Flashes a sketch to a board to obtain the firmware version used by the board",
+		Example: "" +
+			"  " + os.Args[0] + " firmware get-version --fqbn arduino:samd:mkr1000 --address COM10\n" +
+			"  " + os.Args[0] + " firmware get-version -b arduino:samd:mkr1000 -a COM10\n",
+		Args: cobra.NoArgs,
+		Run:  runGetVersion,
+	}
+	commonFlags.AddToCommand(command)
+	return command
+}
+
+func runGetVersion(cmd *cobra.Command, args []string) {
+
+	packageIndex, firmwareIndex := common.InitIndexes()
+	common.CheckFlags(commonFlags.Fqbn, commonFlags.Address)
+	board := common.GetBoard(firmwareIndex, commonFlags.Fqbn)
+	uploadToolDir := common.GetUploadToolDir(packageIndex, board)
+
+	versionSketchPath, err := download.DownloadSketch(board.VersionSketch)
+	if err != nil {
+		feedback.Errorf("Error downloading loader sketch from %s: %s", board.LoaderSketch.URL, err)
+		os.Exit(errorcodes.ErrGeneric)
+	}
+	logrus.Debugf("version sketch downloaded in %s", versionSketchPath.String())
+
+	versionSketch := strings.ReplaceAll(versionSketchPath.String(), versionSketchPath.Ext(), "")
+
+	programmerOut, programmerErr, err := common.FlashSketch(board, versionSketch, uploadToolDir, commonFlags.Address)
+	if err != nil {
+		feedback.Error(err)
+		os.Exit(errorcodes.ErrGeneric)
+	}
+
+	// Wait a bit after flashing the sketch for the board to become available again.
+	logrus.Debug("sleeping for 3 sec")
+	time.Sleep(3 * time.Second)
+
+	currentVersion, err := getVersion(board)
+	if err != nil {
+		feedback.Error(err)
+		os.Exit(1)
+	}
+	if feedback.GetFormat() == feedback.Text {
+		feedback.Printf("Firmware version installed: %s", currentVersion)
+	}
+	// Print the results
+	feedback.PrintResult(&flasher.FlashResult{
+		Programmer: (&flasher.ExecOutput{
+			Stdout: programmerOut.String(),
+			Stderr: programmerErr.String(),
+		}),
+		Version: currentVersion,
+	})
+}
+
+func getVersion(board *firmwareindex.IndexBoard) (fwVersion string, err error) {
+
+	// 9600 is the baudrate used in the CheckVersion sketch
+	port, err := flasher.OpenSerial(commonFlags.Address, 9600, 2)
+	if err != nil {
+		feedback.Error(err)
+		os.Exit(errorcodes.ErrGeneric)
+	}
+
+	buff := make([]byte, 200)
+	serialResult := make([]byte, 0)
+	for {
+		n, err := port.Read(buff)
+		if err != nil {
+			log.Fatal(err)
+			break
+		}
+		serialResult = append(serialResult, buff[:n]...)
+		if n == 0 { // exit when done reading from serial
+			break
+		}
+		logrus.Info(string(buff[:n]))
+	}
+	lines := strings.Split(string(serialResult), "\n")
+	for _, line := range lines {
+		if strings.HasPrefix(line, "Firmware version installed: ") {
+			version := strings.Replace(line, "Firmware version installed: ", "", 1)
+			semver := semver.ParseRelaxed(version)
+			return strings.Replace(semver.String(), "\r", "", 1), nil
```

Comment:
I suggest removing this line ending before parsing. Something like this:
```suggestion
			version := strings.TrimSpace(strings.Replace(line, "Firmware version installed: ", "", 1))
			semver := semver.ParseRelaxed(version)
			return semver.String(), nil
```
Otherwise it breaks the parsing. If you set `semver.WarnInvalidVersionWhenParsingRelaxed = true` with the previous code to get some debug output from the semver package, it prints this message:
```
' invalid patch version separator '8
```
(this was with a FW version of 1.4.8)

Reply:
Nice catch Per! Thanks a lot